### DISCUSSION TITLE
How to simulate S-parameters of passive networks?

### DISCUSSION THREAD

**Discussion by fred3991** [SpiceSharp/SpiceSharp#213] [Q&A] [answered]
Hey, everybody, 
I need an example of how to get the S-parameters of a 2-port circuit ?

I found only AC modeling, but I don't quite understand if it is possible to get S-parameters from the result ?
The circuit looks like this

![изображение](https://github.com/SpiceSharp/SpiceSharp/assets/40190095/72b394da-5f4c-47a3-8a9c-a4638daa8196)

**Comment by svenboulanger** [comment, marked as answer]
Hi,

The S-parameters are calculated using waves, so you will have to first calculate those waves using (see [MathWorks](https://nl.mathworks.com/discovery/s-parameter.html) or [Wikipedia](https://en.wikipedia.org/wiki/Scattering_parameters)).

$$a_i = \frac{1}{2} \frac{V_i+Z_0\cdot I_i}{\sqrt{\left|Re(Z_0)\right|}}$$

$$b_i = \frac{1}{2} \frac{V_i - Z_0^*\cdot I_i}{\sqrt{\left|Re(Z_0)\right|}}$$

Or, whatever convention that you are using.
The $V_i$ and $I_i$ are the complex voltages and currents at port $i$. Once you have these 4 waves, you know that they are related using S-parameters like this:

$$
\left(\begin{array}{c}
b_1\\
b_2
\end{array}\right) = \left(
\begin{array}{cc}
S_{11} & S_{12} \\
S_{21} & S_{22}
\end{array}
\right)\cdot\left(
\begin{array}{c}
a_1\\
a_2
\end{array}
\right)
$$

As you can see, there are 4 unknowns, and only 2 equations, which means that you will need to do 2 linearly independent measurements to find all S-parameters.

Looking at your figure, I'm guessing that the "2 linearly independent measurements" are:

- An AC voltage source at the input of magnitude 1, no source at the output.
- An AC voltage source at the output of magnitude 1, no source at the input.

In both setups, the voltage source is in series with the `TermG` resistor $Z_0$. Then they measure the voltage at the input/output of the DUT, and the current through the $Z_0$, and use those to calculate the waves in both cases. By doing so, they guarantee that $a_1=0$ in one experiment, and $a_2=0$ in another, allowing you to find all S-parameters without the need to invert the matrix. For example, if $a_2=0$, then

$$
\left(
\begin{array}{c}
b_1\\
b_2
\end{array}
\right) = \left(
\begin{array}{cc}
S_{11} & S_{12} \\
S_{21} & S_{22}
\end{array}
\right)\cdot \left(
\begin{array}{c}
a_1 \\
0
\end{array}
\right)
$$

which immediately gives you $S_{11} = \frac{b_1}{a_1}$ and $S_{21} = \frac{b_2}{a_1}$.
